src/terminal/parser/stateMachine.cpp

```diff
@@ -532,6 +537,51 @@ void StateMachine::_ActionParam(const wchar_t wch)
     }
 }
 
+// Routine Description:
+// - Triggers the SubParam action to indicate that the state machine should store this character as a part of a sub-parameter
+//   to a control sequence.
+// Arguments:
+// - wch - Character to dispatch.
+// Return Value:
+// - <none>
+void StateMachine::_ActionSubParam(const wchar_t wch)
+{
+    _trace.TraceOnAction(L"SubParam");
+
+    // Once we've reached the sub-parameter limit, additional sub-parameters are ignored.
+    if (!_subParameterLimitReached)
+    {
+        // On a delimiter, increase the number of sub-params we've seen.
+        // "Empty" sub-params should still count as a sub-param -
+        //      eg "\x1b[38:2::m" should be three sub-params
+        if (_isSubParameterDelimiter(wch))
+        {
+            // If we receive a delimiter after we've already accumulated the
+            // maximum allowed sub-parameters, then we need to set a flag to
+            // indicate that further sub-parameter characters should be ignored.
+            if (_subParameters.size() >= MAX_SUBPARAMETER_COUNT)
+            {
+                _subParameterLimitReached = true;
+            }

```

Comment:
It just occurred to me that it might be easier if you just set the `_parameterLimitReached` flag here as well. That should solve the problem without having to make major changes to your implementation.

So if someone happens to use 32 sub parameters as part of the first parameter, they won't get any additional parameters accepted at all. Dropping excessive parameters is fine - I just don't want them to be misinterpreted.

Reply:
Let says you have an `SGR` sequence with these three parameters:

* `38:2::128:0:128::::::::::::::::::::::::::::`
* `4:0`
* `5`

Ideally all three would be executed, so you get magenta text, underline off, and blinking.

However, I wouldn't mind if we chose to ignore the second two, because the first one exceeded our global parameter limit. It's expected that a terminal might drop excessive parameters, so behavior like this would not be particularly surprising.

Dropping the second one and still executing the third seems a bit weird though. As a user that would look like a bug to me, because it's not at all obvious why that is happening. I don't feel that strongly about it, though, especially since this is never likely to happen in practice.

The only thing I _am_ strongly opposed to is having the second parameter interpreted as `4` instead of `4:0`. So as long as we don't do that, I'm happy.

Reply:
> Ideally all three would be executed, so you get magenta text, underline off, and blinking.

Since these properties are isolated from each other, I'd expect them to work when one of them is not in its standard format (this is already something which application should be blamed for).

I guess, I'd pick a terminal that can still parse `5` (and anything it's capable of, safely) when an application is throwing a bunch of `38:2::128:0:128::::::::::::::::::::::::::::` at it 😄 . It might make our users think, WT is able to parse at least something from a broken `SGR`/`DECCARA`/`DECRARA` sequence its receiving. But again, this is just my opinion 😅 .

It makes me wonder, should we block further parameter only because we couldn't handle sub parameters of a different parameter than we're currently on? We allow up to 32 parameters in a sequence, however, if we starts blocking them, it can come down to 1, if a malformed sub parameters list is received.

> I don't feel that strongly about it, though, especially since this is never likely to happen in practice.

Same. 👍

Reply:
> Since these properties are isolated from each other, I'd expect them to work when one of them is not in its standard format

But in this case _all_ of them are technically valid. It's only our internal implementation that is restricting how many of them we can parse. And that's fine, because the standard allows for terminals to drop parameters past a certain limit, but the expectation is that you're going to keep the first N parameters - we can't just decide we're going to keep the first and the third, but drop the second one.

Reply:
> but the expectation is that you're going to keep the first N parameters - we can't just decide we're going to keep the first and the third, but drop the second one.

That seems reasonable to me, let's refactor the code to behave this way. 👍

Reply:
I guess this is addressed now. Lmk if it needs changes.